wip-Gorka-Irazoqui-ARS-merkelization.md

```diff
@@ -0,0 +1,173 @@
+
+## Specification
+### Forming a ARS with public keys in G2
+The current implementation of BLS signatures in the node assumes that the verification to be done in Ethereum is *e(S, G<sub>2</sub>) = e(H(M), P)*, where:
+
+- *S* is an individual/aggregated signature in G1
+- *G<sub>2</sub>* is the generator in the field G2
+- *H(M)* is the output of a hash to point function that transforms a message *M* and outputs a point in G1.
+- *P* is an individual/aggregated public key in G2.
+
+Therefore, the public keys to be aggregated are those in G2, and that is exactly the reason why the public keys in G2 need to be committed alongside eligibility proofs. With those public keys in G2, the ARS is composed of the public keys of ARS members in G2.
+
+Each of the public keys is composed of 65 bytes in compressed form, and 128 bytes in uncompressed form. Now the question that follows is, should we use the compressed or the uncompressed form to from the merkle tree? To answer the question, we should take a look at how we can calculate the uncompressed from the compressed form. In the case of G2, the uncompressed form calculation involves complex operations in G2. Note that these public keys will need to be provided to the smart contract verifying the ARS membership. Thus, for the moment we decided that such computations are too complex to perform in the Ethereum EVM.
+
+For the aforementioned reasons, the public keys in G2 will need to be merkelized in **uncompressed form**. These keys can still be passed and committed in compressed form by the ARS members, but need to be uncompressed prior to composing the ARS merkle root.
+
+### Order of the leaves
+
+Two possible ordering scenarios are considered for the leaves before computing the merkle root. These options are:
+
+- 1. **ARS ordered by reputation**: allows us to weight votes from those members being more reputed and to standarize the ordering of the members regardless of the curve/public key  associated with it.
+
+- 2. **Public key based order**: this option allows to perform non-membership proofs. However, in the presence of several curves such a feature would require to have a different root per supported curve.
+
+For the case of the bridge we choose to perform a **reputation based ordering**. The reasons are simple: we do not expect the merkle tree to grow to a point up to which non-membership proofs can give us a clear advantage by, e.g., using optimistic rooll-up schemes. In case of tie in reputation, the keys are ordered by pkh. Note that this has little relevancy with respect to the bridge, as long as all nodes have common ordering instructions.
```

Comment:
```suggestion
For the case of the bridge it's been chosen to perform a **reputation based ordering**. The reasons are simple: it is not expected that the merkle tree to grows to a point up to which non-membership proofs can give us a clear advantage by, e.g., using optimistic rooll-up schemes. In case of tie in reputation, the keys are ordered by pkh. Note that this has little relevancy with respect to the bridge, as long as all nodes have common ordering instructions.
```